akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala

```diff
@@ -563,17 +553,27 @@ private[remote] class ArteryTransport(_system: ExtendedActorSystem, _provider: R
               case ActorRefCompressionAdvertisement(from, table) ⇒
                 log.debug("Incoming ActorRef compression advertisement from [{}], table: [{}]", from, table)
                 val a = association(from.address)
-                a.outboundCompression.applyActorRefCompressionTable(table)
-                a.sendControl(ActorRefCompressionAdvertisementAck(localAddress, table.version))
-                system.eventStream.publish(Events.ReceivedActorRefCompressionTable(from, table))
+                // make sure uid is same for active association
+                if (a.associationState.uniqueRemoteAddressValue().exists(_ == from)) {
```

Comment:
There is still one thing I'm uneasy about. I would like to be able to have a check in the `Decoder` for an incoming message that the used compression table for the incoming message was actually created by current incarnation of the destination `ActorSystem`. That would require including the destination uid in each message, and we probably don't want that overhead.

I think it is implicitly taken care of by the handshake and that we reject messages from unknown origin, but using wrong table would be catastrophe (well, not good any way).

I'll take another look at the handshake and see if we can ensure that the outgoing compression table is really cleared before the handshake is completed.

I was also thinking if we can make the compression table version number more unique for each incarnation, instead of not always starting on 1. Might be harder to debug though.